consumer/consumererror/statuserrors.go

```diff
@@ -0,0 +1,87 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"fmt"
+
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// StatusError contains either an HTTP or gRPC status code
+// resulting from a failed request.
+// It allows retrieving a type of error code, converting
+// between the status codes supported by each transport if necessary.
+//
+// It should be created with NewHTTPStatus or NewGRPCStatus.
+type StatusError struct {
+	error
+	httpStatus *int
+	grpcStatus *status.Status
+}
+
+var _ error = &StatusError{}
```

Comment:
Do we want our errors to use value receivers or pointer receivers? Our current errors use value receivers, but it sounds like there may be some arguments for using pointer receivers. The Go devs also appear to use pointer receivers for custom errors: https://go.dev/blog/all

I can see arguments for both sides.

Reply:
`PermanentError` already uses values receivers and I think we should be consistent.  If we want to use pointer receivers we should update that struct.

Do you have a direct link to the go dev custom error blog?

Reply:
>`PermanentError` already uses values receivers and I think we should be consistent. If we want to use pointer receivers we should update that struct.

Agreed, however that would be a breaking change, so I want to make it after this PR. I would also like to document our preference in a style guide. I checked both in this repo and in contrib and we're not consistent. I don't think it's as much of an issue for errors that aren't exported outside their modules, but I think we should at least make a recommendation.

>Do you have a direct link to the go dev custom error blog?

My bad, I copied the wrong link. I've edited my comment, please see the link that is there now.

Reply:
@evan-bradley can you open an issue to capture that we need to make a decision on value vs pointer?

Reply:
After our offline discussion on 2024-04-17 and digging into it a bit deeper, I think we can follow the [official Golang recommendations](https://go.dev/doc/faq#methods_on_values_or_pointers) on value vs. pointer receivers and use value receivers for these objects.

My initial concern was that using value receivers would introduce an issue where doing `err.(*StatusErr)` would cause the value object to be "upgraded" to a pointer and would hide the type mismatch, but I don't think this is an issue since we will suggest using `errors.As`. So there's really no exception to the standard rule here.

I don't think we need to make our own decision on this considering errors shouldn't meet the requirements for pointer receivers in most cases. @TylerHelmuth if you feel differently, I'll open an issue and we can discuss there.